### DISCUSSION TITLE
Merge commits contents can differ between https:...commitid, .patch, and .diff URLs

### DISCUSSION THREAD

**Discussion by hlein** [community/community#56522] [Repositories]
### Select Topic Area

Bug

### Body

This seems like a harmless UI behavior/bug at first, but I can come up with (contrived) scenarios where it has a security impact.

There is a difference between how the GitHub web UI renders merge commits, vs how it publishes raw (`.patch`) versions of that same commit.

See for example:

https://github.com/sysstat/sysstat/commit/6f8dc568e6ab072bb8205b732f04e685bf9237c0

vs:

https://github.com/sysstat/sysstat/commit/6f8dc568e6ab072bb8205b732f04e685bf9237c0.patch

Interestingly, https://github.com/sysstat/sysstat/commit/6f8dc568e6ab072bb8205b732f04e685bf9237c0.diff matches the web UI.

Looking into the commit log locally, I believe what happened and what GH UI is doing is:

1) commit `954ff2e2673cef48f0ed44668c466eab041db387` was made in a fork and a PR was issued, but that patch was slightly incorrect.
2) upstream merged that in merge commit `6f8dc568e6ab072bb8205b732f04e685bf9237c0`, but actually modified it before landing to fix the wrongness.
3) GitHub web UI, when asked to show the changes for `6f8dc568e6ab072bb8205b732f04e685bf9237c0`, shows the actual real/landed version.
4) When downloading `6f8dc568e6ab072bb8205b732f04e685bf9237c0.diff`, it also shows the actual real/landed version.
5) However, when asked for `6f8dc568e6ab072bb8205b732f04e685bf9237c0.patch`, it returns the contents from the commit in the fork that PR came from, commit `954ff2e2673cef48f0ed44668c466eab041db387`

Note, when you do a `git log -p` you see no contents in merge commits, you must do e.g.  `git show -m 6f8dc568e6ab072bb8205b732f04e685bf9237c0` and then you'll see the modified/corrected version.

So I think the web UI and `.diff` are behaving like (basically) `git show` behavior, but `.patch` fetching is not.

Now, how can this have a security impact? Consider a scenario such as:

1) Mallory finds a repo that she knows some users fetch individual fixes as `.patch` files from.
2) Mallory gains temporary malicious write access to that repository.
3) Mallory submits a PR containing malicious code, backdoor, etc. under another GH account.
4) Mallory uses privileged access to merge that PR but in the process, squashes out anything malicious.
5) Bob reads diffs through the web UI, everything looks fine.
6) Bob downloads the `.patch` version and deploys it.
7) profit

(I *said* it was contrived.)

**Comment by hlein** [comment]
Note that from the terse description for CVE-2023-23764, it sounds like the same or similar issue as this. That was reported to hackerone about GitHub Enterprise (before I created this, but it was embargoed until today): https://hackerone.com/reports/1938106

So GitHub recognized this or something very similar as a security bug, paid a bounty for it, reportedly resolved it, but only in GitHub Enterprise. The proof of concept links above for github.com still work.

Looks like @s0larized worked on handling the hackerone report.

**Comment by ManishParashar0** [comment]
The scenario you've described illustrates a potential security risk that arises from inconsistencies in how GitHub handles the display and retrieval of changes in merge commits. When users view merge commits through the GitHub web UI or download them as a .diff file, they see the correct, modified version that was landed. However, when users download the merge commit as a .patch file, they receive the contents from the original commit in the fork.

This discrepancy could be exploited by malicious actors like Mallory in the following way:

1. Mallory gains temporary write access to a repository where users typically fetch individual fixes as .patch files.

2. Mallory submits a malicious PR under another GitHub account. The malicious code or backdoor is intended to be part of the commit that gets squashed during the merge process.

3. Using privileged access, Mallory merges the PR but modifies it before landing, squashing out any malicious content. The GitHub web UI and .diff file display the correct, safe version.

4. However, when users download the merge commit as a .patch file, they unknowingly receive the contents from the original, malicious commit in the fork.

5. Unsuspecting users may deploy the .patch file, unintentionally introducing the malicious code or backdoor into their systems.

This is a contrived scenario, but it highlights a potential security concern. Users who rely on .patch files to fetch and apply changes could be exposed to risks if there are inconsistencies in how GitHub handles the retrieval of the correct content for merge commits.

To mitigate such risks, GitHub should ensure that the behavior for downloading .patch files aligns with the actual merged version, just like how the web UI and .diff files display the correct content. Consistency in how merge commits are handled across different retrieval methods is crucial for maintaining the integrity and security of the codebase.

If you suspect or discover any such security vulnerabilities or inconsistencies, it's essential to responsibly disclose them to GitHub's security team to address and fix the issues promptly. GitHub has a security vulnerability reporting page that you can use to report such incidents responsibly.

**Reply by hlein** [reply]
Yes, you've pretty much restated the scenario I described.

Exploitation of this is pretty far-fetched, so I don't consider it particularly urgent or sensitive. (In a typical project, Mallory has already won if she gets repo write access, since almost nobody reviews code from repos they clone. Or even require signed commits, etc.)

I spent at least 30 seconds looking for a good way to report it privately (to be clear: I acknowledge I did not try very hard). I found it surprisingly difficult. None of these get to something useful:
* Google for: github security <- marketing page
* github.com/security <- marketing page
* Google for: report github vulnerability <- talks about vuln reports for repos hosted on github, not github vulns
* Google for: github security advisory <- same as above
* the "Contact Github" footer link <- only thing security-related is another marketing page
* support.github.com <- same as above
* the "Contact us" link on the support page <- nothing specific; closest thing was this community discussion
* searched various PGP keyservers for a `security@github.com` key, none found

I could have blindly tried to email `security@` or some others and see what happens, but no. If I spent more time than that running down private contact methods for every bug I find that *might* have interesting security ramifications, I'd never get any work done.

Now, I thought to check https://github.com/.well-known/security.txt and that works - and points at hackerone.com to submit a report. But I don't have an account there and GitHub is trying to create chores for me, no thanks.
